libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,269 @@
+    Generic,
+    Any,
+    Literal,
+    Dict,
+)
+
+from langchain_core.load.serializable import Serializable
+from langchain_core.runnables import RunnableConfig
+
+
+class IndexableData(
+    ABC, Serializable
+):  # Introduce new base class for content. Name OK?
+    id: str
+    metadata: Optional[dict]  # In a more extreme form we could remove
+
+
+class Document(IndexableData):  # Re-use existing Document model in langchain_core
+    content: str
+
+
+class Blob(IndexableData):  # Re-use existing Blob model in langchain_core
+    """Use for images / audio / video"""
+
+    data: bytes
+    encoding: str  # base64, utf-8, URI, etc.
+    mimetype: str  # image/png, text/plain, etc.
+
+
+# This is likely not needed during indexing right now
+# class MultiContent(Content):
+#     """Group multiple pieces of content together"""
+#     contents: List[Content]
+#     metadata: dict
+
+
+T = TypeVar("T", bound=IndexableData)
+
+
+class UpsertResponse(TypedDict):
+    """An indexing result."""
+
+    failed: Sequence[str]
+    indexed: Sequence[str]
+
+
+class DeleteResponse(TypedDict):
+    """A response to a delete request."""
+
+    num_deleted: NotRequired[int]
+    num_failed: NotRequired[int]
+    failed: NotRequired[Sequence[str]]
+    deleted: NotRequired[Sequence[str]]
+
+
+class Sort(TypedDict):
+    """A sort object."""
+
+    field: str
+    order: NotRequired[Literal["asc", "desc"]]  # Assume asc by default
+
+
+class FilterQuery(TypedDict):
+    """Query for an item.
+
+    This enables querying for an item using similarity search +
+    standard operations on the relational data associated with the item.
+    """
+
+    filter: Optional[dict]  # We'll need to type this # <-- Could stand for raw filter?
+    limit: Optional[int]  # equivalent to top-k right now
+    offset: Optional[int]  # e.g., interpret as page offset (maybe a better name?)
+    # e.g., browse through documents by publication date
+    sort: Optional[List[Sort]]  # Optional multiple sort fields
+
+
+C = TypeVar("C")
+
+
+Vector = List[float]
+
+
+class BaseIndex(Generic[T]):
+    """An index represent a collection of items that can be queried.
+
+    The index is responsible for:
+
+    1. Storing the content
+    2. Allowing retrieval of the content by id
+    3. Supporting queries against the metadata associated with the content
+
+    The index is **NOT** responsible for:
+
+    1. Support search queries against the content.
+
+    The types of supported queries seems very dependent on the actual
+    """
+
+    @abc.abstractmethod
+    def upsert(self, data: Iterable[T], /, **kwargs: Any) -> Iterable[UpsertResponse]:
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Iterable[str],  # Sequence or Iterable
+        /,
+    ) -> Iterable[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],  # Sequence or Iterable
```

Comment:
why would get and delete be diff (re sequence or iterable)

Reply:
expressing my uncertainty -- I'm inclined to use Sequence for both